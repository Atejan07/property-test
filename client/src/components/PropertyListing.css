.listingContainer {
  display: flex;
  justify-content: flex-start;
  flex-wrap: wrap;
  margin: 0 4rem;

  @media (max-width: 1000px) {
    flex-direction: column;
    align-items: center;
    margin: 0 2rem;
  }
}

.propertyItem {
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  width: 300px;
  overflow: hidden;
  border-radius: 8px;
  flex-basis: calc(33.33% - 4rem);
  margin: 2rem;
  max-width: 30vw;
  background-color: var(--secondary);

  @media (max-width: 1000px) {
    max-width: 90vw;
    margin: 2rem 0;
    width: 600px;
  }

  @media (max-width: 500px) {
    width: auto;
  }
}

.photoContainer {
  display: flex;
  justify-content: center;
  height: 200px;
  background-color: var(--tertiary);
}

.propertyItem img {
  max-width: 100%;
  object-fit: cover;
}

.propertyDetails {
  padding: 0.5rem 1.5rem;
}

.propertyDetails h3 {
  color: var(--primary);
}

.propertyDetails h2 {
  color: var(--tertiary);
}

.propertyDetails p {
  font-weight: 100;
  text-align: justify;
}

.propertyDetails #address {
  font-weight: bold;
}
