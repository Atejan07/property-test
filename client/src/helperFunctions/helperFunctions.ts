import { Error, Property } from '../types/types';

export function validatePostalCodeUK(postalCode: string) {
  const ukPostalCodeRegex =
    /^(([A-Z][0-9]{1,2})|(([A-Z][A-HJ-Y][0-9]{1,2})|(([A-Z][0-9][A-Z])|([A-Z][A-HJ-Y][0-9]?[A-Z])))) [0-9][A-Z]{2}$/i;

  if (!ukPostalCodeRegex.test(postalCode)) {
    return 'Please provide a valid UK postal code format (e.g. EC4V 4BE).';
  }
  return undefined;
}

export function validateForm(values: Property) {
  const errors: Error = {};

  // Required field validation
  const requiredFields = {
    houseNumber: 'house number',
    streetName: 'street name',
    city: 'city',
    country: 'country',
    postalCode: 'postal code',
    description: 'description',
    photo: 'property photo',
    price: 'price',
  };
  Object.entries(requiredFields).forEach((field) => {
    if (!values[field[0] as keyof Property]) {
      errors[field[0] as keyof Error] = `Please, provide a ${field[1]}.`;
    }
  });

  // Numeric field validation
  const alphanumericRegex = /^[a-zA-Z0-9\s-]+$/;

  if (
    !values.houseNumber ||
    isNaN(parseFloat(values.houseNumber)) ||
    !alphanumericRegex.test(values.houseNumber)
  ) {
    errors.houseNumber = 'Please provide a valid house number (e.g. 22).';
  }

  if (
    !values.price ||
    isNaN(parseFloat(values.price)) ||
    !/^\d{5,}$/.test(values.price)
  ) {
    errors.price =
      'Please ensure that the price consists of a minimum of 5 digits and contains only numerical values (e.g. 150000).';
  }

  if (values.bedrooms <= 0) {
    errors.bedrooms =
      'Please provide a valid number of bedrooms (should be at least 1).';
  }
  if (values.bathrooms <= 0) {
    errors.bathrooms =
      'Please provide a valid number of bathrooms (should be at least 1).';
  }

  // Alphabetic field validation
  const alphabeticRegex = /^[a-zA-Z\s-]+$/;
  if (!alphabeticRegex.test(values.streetName)) {
    errors.streetName =
      'Please provide a valid street name (e.g. Oxford Street).';
  }

  if (!alphabeticRegex.test(values.city)) {
    errors.city = 'Please provide a valid city name (e.g. London).';
  }

  if (!alphabeticRegex.test(values.country)) {
    errors.country =
      'Please provide a valid country name (e.g. United Kingdom).';
  }

  // Postal code validation
  if (values.country || values.postalCode) {
    if (values.country === 'United Kingdom') {
      const ukPostalCodeError = validatePostalCodeUK(values.postalCode);
      if (ukPostalCodeError) {
        errors.postalCode = ukPostalCodeError;
      }
    }
  }
  return errors;
}

export function formatPrice(price: string) {
  const formattedPrice = parseFloat(price).toLocaleString('en-GB', {
    style: 'currency',
    currency: 'GBP',
  });
  return formattedPrice;
}
