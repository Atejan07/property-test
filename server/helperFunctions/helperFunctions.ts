import { properties } from '../db/db';
import { Property } from '../types/types';

export function checkDuplicateProperty(property: Property) {
  return properties.some((prop) => {
    return (
      prop.houseNumber === property.houseNumber &&
      prop.streetName === property.streetName &&
      prop.city === property.city &&
      prop.country === property.country &&
      prop.postalCode === property.postalCode &&
      prop.propertyType === property.propertyType
    );
  });
}

export function checkEmptyProperty(property: Property): boolean {
  return (
    !property.houseNumber ||
    !property.streetName ||
    !property.city ||
    !property.country ||
    !property.postalCode ||
    !property.propertyType
  );
}
