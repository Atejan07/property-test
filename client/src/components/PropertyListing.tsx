import { formatPrice } from '../helperFunctions/helperFunctions';
import { Property } from '../types/types';
import './PropertyListing.css';

const defaultPhoto =
  'https://images.pexels.com/photos/186077/pexels-photo-186077.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2';

interface PropertyListingProps {
  properties: Property[];
}

function PropertyListing({ properties }: PropertyListingProps) {
  return (
    <>
      <h1>Properties for sale</h1>
      <div className='listingContainer'>
        {properties.length > 0 ? (
          properties.map((property) => (
            <div
              key={property.id}
              className='propertyItem'
              data-testid='propertyItem'
            >
              <div className='photoContainer'>
                <img
                  src={property.photo || defaultPhoto}
                  alt={property.photo}
                />
              </div>
              <div className='propertyDetails'>
                <h3>{property.bedrooms} bedrooms for sale</h3>
                <p>{property.propertyType}</p>
                <h2>{formatPrice(property.price)}</h2>
                <p>{property.description}</p>
                <p id='address'>
                  {property.streetName}, {property.city}, {property.postalCode}
                </p>
              </div>
            </div>
          ))
        ) : (
          <div className='listingContainer'>
            <h2>Sorry, there are no properties today.</h2>
          </div>
        )}
      </div>
    </>
  );
}

export default PropertyListing;
