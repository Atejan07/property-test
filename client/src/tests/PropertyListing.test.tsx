import { render, waitFor, screen } from '@testing-library/react';
import PropertyListing from '../components/PropertyListing';
import { mockProperties } from './mocks';
import { formatPrice } from '../helperFunctions/helperFunctions';

test('renders correctly', () => {
  const { container } = render(<PropertyListing properties={[]} />);
  expect(container).toMatchSnapshot();
});

test('has the correct heading', () => {
  render(<PropertyListing properties={[]} />);
  expect(
    screen.getByRole('heading', { name: 'Properties for sale' })
  ).toBeInTheDocument();
});

test('renders propertyType, description and price', async () => {
  render(<PropertyListing properties={mockProperties} />);
  await waitFor(() => {
    for (const property of mockProperties) {
      expect(screen.getByText(property.propertyType)).toBeInTheDocument();
      expect(screen.getByText(property.description)).toBeInTheDocument();
      expect(screen.getByText(formatPrice(property.price))).toBeInTheDocument();
    }
  });
});

test('renders a message if there is no properties', async () => {
  render(<PropertyListing properties={[]} />);
  expect(
    screen.getByText('Sorry, there are no properties today.')
  ).toBeInTheDocument();
});

test('doesnt render houseNumber and streetName', async () => {
  render(<PropertyListing properties={[]} />);
  await waitFor(() => {
    for (const property of mockProperties) {
      expect(screen.queryByText(property.houseNumber)).not.toBeInTheDocument();
      expect(screen.queryByText(property.streetName)).not.toBeInTheDocument();
    }
  });
});

test('displays correct number of property items', () => {
  render(<PropertyListing properties={mockProperties} />);
  const propertyItems = screen.getAllByTestId('propertyItem');
  expect(propertyItems).toHaveLength(mockProperties.length);
});

test('renders single property correctly', () => {
  const singleProperty = [mockProperties[0]];
  render(<PropertyListing properties={singleProperty} />);

  const propertyElement = screen.getByTestId('propertyItem');
  expect(propertyElement).toBeInTheDocument();
});
