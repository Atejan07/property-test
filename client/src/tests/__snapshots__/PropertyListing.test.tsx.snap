// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <h1
    id="firstTitle"
  >
    Properties for sale
  </h1>
  <div
    class="listingContainer"
  >
    <div
      class="listingContainer"
    >
      <h2>
        Sorry, there are no properties today.
      </h2>
    </div>
  </div>
</div>
`;
